File: packages/scheduling-engine/src/scheduler/dispatch-rules.ts
// =====================================================================
//  INCOMPOL PLAN -- Dispatch Rules & Group Sorting
//  Comparators for EDD, CR, WSPT, SPT dispatch rules.
//  Tool merging logic to reduce redundant setups.
//  Supply boost priority override.
//
//  Pure functions -- no React, no side effects.
// =====================================================================

import { DAY_CAP, MAX_EDD_GAP } from '../constants.js';
import type { DispatchRule } from '../types/kpis.js';
import type { ATCSParams } from './atcs-dispatch.js';
import { atcsPriority, computeATCSAverages, DEFAULT_ATCS_PARAMS } from './atcs-dispatch.js';
import type { ToolGroup } from './demand-grouper.js';

// ── Supply boost extraction ─────────────────────────────────────────

/** Get the maximum supply boost for a tool group */
function maxBoost(g: ToolGroup, supplyBoosts?: Map<string, { boost: number }>): number {
  if (!supplyBoosts) return 0;
  return Math.max(0, ...g.skus.map((sk) => supplyBoosts.get(sk.opId)?.boost ?? 0));
}

// ── Dispatch comparators ────────────────────────────────────────────

/**
 * Creates a group comparator function for the specified dispatch rule.
 *
 * Rules:
 * - **EDD** (Earliest Due Date): lower edd first, tiebreak by larger production
 * - **CR** (Critical Ratio): (time remaining) / (processing time), lower = more urgent
 * - **WSPT** (Weighted Shortest Processing Time): weight/time descending
 * - **SPT** (Shortest Processing Time): ascending production time
 *
 * Supply boost always takes primary priority (higher boost = schedule first).
 *
 * @param rule          - Dispatch rule to use
 * @param supplyBoosts  - Optional supply boost map
 * @returns Comparator function for Array.sort()
 */
export function createGroupComparator(
  rule: DispatchRule,
  supplyBoosts?: Map<string, { boost: number }>,
  atcsContext?: { avgProdMin: number; avgSetupMin: number; params: ATCSParams },
): (a: ToolGroup, b: ToolGroup) => number {
  return (a: ToolGroup, b: ToolGroup): number => {
    // Supply boost takes priority -- higher boost = schedule first
    const ba = maxBoost(a, supplyBoosts);
    const bb = maxBoost(b, supplyBoosts);
    if (ba !== bb) return bb - ba;

    switch (rule) {
      case 'CR': {
        // Critical Ratio: (time remaining) / (processing time). Lower = more urgent.
        // Uses DAY_CAP because totalProdMin already has OEE baked in
        // from demand-grouper: prodMin = (qty / pH) * 60 / OEE
        const crA = a.edd <= 0 ? 0 : a.edd / Math.max(a.totalProdMin / DAY_CAP, 0.01);
        const crB = b.edd <= 0 ? 0 : b.edd / Math.max(b.totalProdMin / DAY_CAP, 0.01);
        return crA !== crB ? crA - crB : b.totalProdMin - a.totalProdMin;
      }
      case 'WSPT': {
        // Weighted Shortest Processing Time: weight/time descending (higher ratio = schedule first)
        const totalA = a.skus.reduce((s, sk) => s + sk.totalQty, 0);
        const totalB = b.skus.reduce((s, sk) => s + sk.totalQty, 0);
        const ratioA = totalA / Math.max(a.totalProdMin, 1);
        const ratioB = totalB / Math.max(b.totalProdMin, 1);
        return ratioB - ratioA;
      }
      case 'SPT':
        // Shortest Processing Time: ascending production time
        return a.totalProdMin !== b.totalProdMin ? a.totalProdMin - b.totalProdMin : a.edd - b.edd;
      case 'ATCS': {
        // ATCS: higher priority index = schedule first
        const ctx = atcsContext ?? {
          avgProdMin: 100,
          avgSetupMin: 30,
          params: DEFAULT_ATCS_PARAMS,
        };
        // Static comparator: no previous-tool context, so previousToolId = null
        const prioA = atcsPriority(a, null, ctx.params, ctx.avgProdMin, ctx.avgSetupMin);
        const prioB = atcsPriority(b, null, ctx.params, ctx.avgProdMin, ctx.avgSetupMin);
        return prioB - prioA;
      }
      default: // EDD
        return a.edd !== b.edd ? a.edd - b.edd : b.totalProdMin - a.totalProdMin;
    }
  };
}

// ── Tool merging ────────────────────────────────────────────────────

/**
 * Merge consecutive tool-groups for the same tool to reduce redundant setups.
 *
 * After EDD sort, Tool-A-bucket1 then Tool-B-bucket then Tool-A-bucket2
 * would cause 2 setups for Tool-A. This pulls Tool-A-bucket2 right after
 * bucket1, but ONLY if the EDD gap <= MAX_EDD_GAP to avoid delaying
 * intermediate deliveries.
 *
 * @param groups      - Sorted tool groups for one machine
 * @param maxEddGap   - Max days between buckets to allow merging
 * @returns Merged group list with reduced setup transitions
 */
export function mergeConsecutiveTools(
  groups: ToolGroup[],
  maxEddGap: number = MAX_EDD_GAP,
): ToolGroup[] {
  const merged: ToolGroup[] = [];
  const used = new Set<number>();

  for (let i = 0; i < groups.length; i++) {
    if (used.has(i)) continue;
    used.add(i);
    merged.push(groups[i]);

    // Pull forward subsequent buckets for the same tool (if EDD gap is small enough)
    let lastEdd = groups[i].edd;
    for (let j = i + 1; j < groups.length; j++) {
      if (used.has(j)) continue;
      if (groups[j].toolId === groups[i].toolId && groups[j].edd - lastEdd <= maxEddGap) {
        used.add(j);
        merged.push(groups[j]);
        lastEdd = groups[j].edd;
      }
    }
  }

  return merged;
}

/**
 * G5: Material-part merging -- group tools that share the same MP consecutively.
 * Reduces material changeovers (same coil/sheet can serve multiple tools).
 * Only reorders within a small EDD window to avoid delaying urgent deliveries.
 *
 * @param groups      - Tool-merged group list
 * @param maxEddGap   - Max EDD gap for material merging
 * @returns Groups reordered to minimize material changeovers
 */
export function mergeMaterialParts(
  groups: ToolGroup[],
  maxEddGap: number = MAX_EDD_GAP,
): ToolGroup[] {
  const mpMerged: ToolGroup[] = [];
  const mpUsed = new Set<number>();

  for (let i = 0; i < groups.length; i++) {
    if (mpUsed.has(i)) continue;
    mpUsed.add(i);
    mpMerged.push(groups[i]);

    const mp = groups[i].tool.mp;
    if (!mp) continue;

    // Pull forward groups with same MP (if EDD gap is small)
    for (let j = i + 1; j < groups.length; j++) {
      if (mpUsed.has(j)) continue;
      if (groups[j].tool.mp === mp && groups[j].edd - groups[i].edd <= maxEddGap) {
        mpUsed.add(j);
        mpMerged.push(groups[j]);
      }
    }
  }

  return mpMerged;
}

// ── SKU intra-group sort ────────────────────────────────────────────

/**
 * Sort SKUs within each tool group by urgency:
 * 1. Backlog (atr > 0) first
 * 2. Zero-stock items with lot economic first
 * 3. Largest quantity first (tiebreaker)
 *
 * @param groups - Tool groups (mutated in place)
 */
export function sortSkusWithinGroups(groups: ToolGroup[]): void {
  for (const g of groups) {
    g.skus.sort((a, b) => {
      if (a.atr > 0 && b.atr === 0) return -1;
      if (b.atr > 0 && a.atr === 0) return 1;
      if (a.stk === 0 && a.lt > 0 && !(b.stk === 0 && b.lt > 0)) return -1;
      if (b.stk === 0 && b.lt > 0 && !(a.stk === 0 && a.lt > 0)) return 1;
      return b.totalQty - a.totalQty;
    });
  }
}

// ── Full sort + merge pipeline ──────────────────────────────────────

/**
 * Apply the complete dispatch pipeline to groups for one machine:
 * 1. Sort by dispatch rule (with supply boost override)
 * 2. Merge consecutive same-tool groups
 * 3. Merge same-material-part groups
 * 4. Sort SKUs within each group
 *
 * @param groups         - Unsorted tool groups for one machine
 * @param rule           - Dispatch rule to apply
 * @param supplyBoosts   - Optional supply boost map
 * @returns Sorted, merged groups ready for scheduling
 */
export function sortAndMergeGroups(
  groups: ToolGroup[],
  rule: DispatchRule,
  supplyBoosts?: Map<string, { boost: number }>,
  atcsParams?: ATCSParams,
): ToolGroup[] {
  // For ATCS, pre-compute averages from the group set (needed by the priority formula)
  const atcsContext =
    rule === 'ATCS'
      ? { ...computeATCSAverages(groups), params: atcsParams ?? DEFAULT_ATCS_PARAMS }
      : undefined;
  const comparator = createGroupComparator(rule, supplyBoosts, atcsContext);
  const sorted = [...groups].sort(comparator);

  const toolMerged = mergeConsecutiveTools(sorted);
  const mpMerged = mergeMaterialParts(toolMerged);

  sortSkusWithinGroups(mpMerged);

  return mpMerged;
}

// ── Machine ordering ────────────────────────────────────────────────

/**
 * Order machines by urgency of their first group.
 * Machines with no-alt tools get priority (can't be rerouted).
 *
 * @param machines      - All machines
 * @param mGroups       - Machine -> sorted groups map
 * @param comparator    - Group comparator function
 * @returns Machines sorted by scheduling urgency
 */
export function orderMachinesByUrgency(
  machines: { id: string }[],
  mGroups: Record<string, ToolGroup[]>,
  comparator: (a: ToolGroup, b: ToolGroup) => number,
): { id: string }[] {
  return machines
    .filter((m) => mGroups[m.id]?.length)
    .sort((a, b) => {
      const urgA = mGroups[a.id]?.[0];
      const urgB = mGroups[b.id]?.[0];
      if (urgA && urgB) {
        const d = comparator(urgA, urgB);
        if (d !== 0) return d;
      }
      // Tiebreaker: machines with no-alt tools get priority (can't be rerouted)
      const aNoAlt = mGroups[a.id]?.some((g) => g.skus.some((s) => !s.hasAlt)) ? 0 : 1;
      const bNoAlt = mGroups[b.id]?.some((g) => g.skus.some((s) => !s.hasAlt)) ? 0 : 1;
      if (aNoAlt !== bNoAlt) return aNoAlt - bNoAlt;
      return a.id.localeCompare(b.id);
    });
}

File: apps/backend/src/domain/scheduler/dispatch_rules.py
"""Dispatch rules — section 3 of the spec.

Five priority rules: ATCS, EDD, CR, SPT, WSPT.
Tool merging and machine urgency ordering.
"""

from __future__ import annotations

import math
from collections.abc import Callable

from .constants import DAY_CAP, K1_VALUES, K2_VALUES, MAX_EDD_GAP
from .types import ToolGroup

# ── Comparator helpers ──


def _supply_boost(g: ToolGroup) -> float:
    """Max supply boost from any bucket in the group (higher = first)."""
    return 0.0  # No supply boosts in base pipeline


def _edd_sort_key(g: ToolGroup) -> tuple[float, float, float]:
    """EDD: smallest edd first, then largest prodMin first."""
    return (-_supply_boost(g), g.edd, -g.total_prod_min)


def _cr_sort_key(g: ToolGroup) -> tuple[float, float, float]:
    """CR: smallest critical ratio first."""
    cr = g.edd / max(g.total_prod_min / DAY_CAP, 0.01)
    return (-_supply_boost(g), cr, -g.total_prod_min)


def _wspt_sort_key(g: ToolGroup) -> tuple[float, float]:
    """WSPT: highest weight/time ratio first."""
    w = 1e6 if g.edd <= 0 else 1.0 / g.edd
    ratio = w / max(g.total_prod_min, 1.0)
    return (-_supply_boost(g), -ratio)


def _spt_sort_key(g: ToolGroup) -> tuple[float, float, float]:
    """SPT: shortest processing time first."""
    return (-_supply_boost(g), g.total_prod_min, g.edd)


def atcs_index(
    total_prod_min: float,
    edd: int,
    setup_min: float,
    k1: float,
    k2: float,
    p_bar: float,
    s_bar: float,
    last_tool: str | None,
    tool_id: str,
) -> float:
    """ATCS priority index I(g). Higher = schedule first."""
    p = max(total_prod_min, 1.0)
    slack = max(edd * DAY_CAP - p, 0.0)
    setup = 0.0 if last_tool == tool_id else setup_min

    term1 = 1.0 / p
    term2 = math.exp(-slack / (k1 * p_bar)) if k1 * p_bar > 0 else 1.0
    term3 = math.exp(-setup / (k2 * s_bar)) if k2 * s_bar > 0 else 1.0

    return term1 * term2 * term3


# ── Sorting ──


def sort_groups(
    groups: list[ToolGroup],
    rule: str = "EDD",
    k1: float = 1.5,
    k2: float = 0.5,
    last_tool: str | None = None,
) -> list[ToolGroup]:
    """Sort ToolGroups by dispatch rule. Returns new sorted list."""
    if rule == "ATCS":
        p_bar, s_bar = _compute_p_bar_s_bar(groups)
        return sorted(
            groups,
            key=lambda g: (
                -atcs_index(
                    g.total_prod_min,
                    g.edd,
                    g.setup_min,
                    k1,
                    k2,
                    p_bar,
                    s_bar,
                    last_tool,
                    g.tool_id,
                )
            ),
        )
    elif rule == "CR":
        return sorted(groups, key=_cr_sort_key)
    elif rule == "WSPT":
        return sorted(groups, key=_wspt_sort_key)
    elif rule == "SPT":
        return sorted(groups, key=_spt_sort_key)
    else:  # EDD (default)
        return sorted(groups, key=_edd_sort_key)


def _compute_p_bar_s_bar(groups: list[ToolGroup]) -> tuple[float, float]:
    """Average processing and setup times for ATCS normalization."""
    if not groups:
        return 1.0, 1.0
    p_bar = sum(g.total_prod_min for g in groups) / len(groups)
    s_bar = sum(g.setup_min for g in groups) / len(groups)
    return max(p_bar, 1.0), max(s_bar, 1.0)


# ── Tool merging ──


def merge_consecutive_tools(groups: list[ToolGroup]) -> list[ToolGroup]:
    """Section 3.2: Merge consecutive groups with same tool and EDD gap ≤ MAX_EDD_GAP."""
    if len(groups) <= 1:
        return groups

    merged: list[ToolGroup] = [groups[0]]
    for g in groups[1:]:
        prev = merged[-1]
        if (
            g.tool_id == prev.tool_id
            and g.machine_id == prev.machine_id
            and abs(g.edd - prev.edd) <= MAX_EDD_GAP
        ):
            # Merge: combine buckets, take max EDD
            prev.buckets.extend(g.buckets)
            prev.total_prod_min += g.total_prod_min
            prev.edd = max(prev.edd, g.edd)
        else:
            merged.append(g)

    return merged


# ── Machine urgency ordering ──


def order_machines_by_urgency(
    machine_groups: dict[str, list[ToolGroup]],
    rule: str = "EDD",
) -> list[str]:
    """Section 3.3: Order machines by their most urgent group."""

    def urgency(mid: str) -> tuple:
        groups = machine_groups[mid]
        if not groups:
            return (float("inf"),)
        best = min(groups, key=lambda g: g.edd)
        return (best.edd, -best.total_prod_min)

    return sorted(machine_groups.keys(), key=urgency)


# ── ATCS grid search ──


def grid_search_atcs(
    machine_groups: dict[str, list[ToolGroup]],
    schedule_fn: Callable[..., float],
) -> tuple[float, float, float]:
    """Section 7: 25 K1×K2 combos. Returns (best_k1, best_k2, best_score)."""
    best_k1, best_k2, best_score = 1.5, 0.5, float("inf")

    for k1 in K1_VALUES:
        for k2 in K2_VALUES:
            score = schedule_fn(k1=k1, k2=k2)
            if score < best_score:
                best_score = score
                best_k1 = k1
                best_k2 = k2

    return best_k1, best_k2, best_score
